### PR TITLE
DataWriter doesn't remove unregistered instances

### PR DESCRIPTION
**Pull request by jrw972**
Problem
-------

Entries in the maps of DataWriterImpl_t and WriteDataContainer
correspond to instances.  Unregistering an instance should remove from
these maps to prevent a memory leak.

Solution
--------

Remove from these maps when instances are unregistered.

### CODE REVIEW

File: dds/DCPS/DataWriterImpl_T.h [hunk 1]
```diff
@@ -146,21 +148,34 @@ DataWriterImpl_T
...
+
+      instance_handle = pos->second;
+      reverse_instance_map_.erase(pos->second);
+      instance_map_.erase(pos);
+    } else {
+      typename ReverseInstanceMap::iterator pos = reverse_instance_map_.find(instance_handle);
+      if (pos == reverse_instance_map_.end()) {
+        return DDS::RETCODE_BAD_PARAMETER;
+      }
+      // TODO: Will this comparison work correctly?
```
Comment (mitza-oci) [inline dds/DCPS/DataWriterImpl_T.h:170]: TODO

**Review by mitza-oci — CHANGES_REQUESTED** [review]

**Review by mitza-oci — APPROVED** [review]

File: dds/DCPS/WriteDataContainer.cpp [hunk 2]
```diff
@@ -1322,57 +1342,48 @@ WriteDataContainer::unregister_all()
...
-    // Get the next iterator before erase the instance handle.
-    PublicationInstanceMapType::iterator it_next = it;
-    ++it_next;
-    writer_->return_handle(it->first);
-    // Remove the instance from the instance list.
-    unbind(instances_, it->first);
-    it = it_next;
+    writer_->return_handle(pos->first);
+    instances_.erase(pos++);
+    ACE_UNUSED_ARG(registered_sample);
```
Comment (iguessthislldo) [inline dds/DCPS/WriteDataContainer.cpp:1385]: I don't think this line is needed since this variable is always used.

**Review by iguessthislldo — CHANGES_REQUESTED** [review]

**Comment by jwillemsen** [comment]
Any explicit unit test possible for this?

**Comment by jrw972** [comment]
> Any explicit unit test possible for this?

Possible yes.  However, I think it's going to take a serious refactoring of DataWriterImpl_T, DataWriterImpl, WriteDataContainer, and the associated transport framework to write it as a unit test and not an integration test.

**Review by iguessthislldo — APPROVED** [review]

**Comment by mitza-oci** [comment]
> Any explicit unit test possible for this?

FooTest3_0 is not exactly a unit test, but it will provide coverage of these code paths

**Pull request merged**